arrow-site/docs/docs/patterns/monads/README.md

```diff
@@ -365,318 +348,239 @@ Subsequently, `addStep` is called two times to transfer to `Conference` and then
+}
 
-```kotlin
-repository
-    .loadSpeaker()
-    .flatMap { speaker -> speaker.nextTalk()
-        .flatMap { talk -> talk.getConference() }
-        .flatMap { conference -> conference.getCity().map { it to speaker } }
-    }
-    .runSync { (city, speaker) -> city.fold(
-        { Logger.logError(it); reservations.cancel() },
-        { reservations.bookFlight(speaker, city) })
-    }
+suspend fun OptionRepository.shipperOfLastOrderOnCurrentAddress(customerId: Int): Option<Shipper> =
+  option {
+    val customer = getCustomer(customerId).bind() 
+    val address = getAddress(customer.addressId).bind()
+    val lastOrder = address.lastOrder.bind()
+    val order = getOrder(lastOrder.id).bind()
+    order.shipper
+  }
 ```
 
-We are back to the familiar structure. Time for some more complications.
+We observe here how in the same way we implemented and used monad comprehensions for nullable types, we now implement the same Monad pattern, this time over Option. Fortunately Arrow already provides all these effect builders for all the data types that support monadic behavior.
 
 ### Abstraction for all Monads
 
 We're going to dispel one common misconception.
 Sometimes the word Monad is used to refer to types like Option, Future, Either, and so on, and that's not correct.
-Those are called datatypes or just types. Let's see the difference!
+Those are called data-types or just types. Let's see the difference!
 
-As you have seen, neither Future nor Option implement Monad directly.
-This is intentional, as you can potentially have several Monad implementations for a single type.
-For example, RxJava's Observable can be chained using `flatMap`, `switchMap`, and `concatMap`, and using each is still a Monad.
+As you have seen, neither A? nor Option implement Monad directly. Their actual `bind` and `just` capabilities are described by implementing the `Effect` interface.
 
-Instead, Arrow specifies that Monad must be implemented by a separate object or class, referred to as the "instance of Monad for type F".
+This is intentional, as you can potentially have several Monad implementations for a single type or even mix non-monadic effects inside `Effect` interfaces in order to provide more expressive DSLs. This is notorious in the actual Arrow implementations where effects such as `Either` can perform different forms of monad bind over multiple types like `Either` and `Validated` using the same idioms.
 
-```kotlin
-object FutureMonad: Monad<ForFuture> {
-    override fun <A> just (instance: A): Future<A> =
-        Future(a)
+Arrow specifies that Monad must be implemented by a separate object or class in terms of the `Effect` interface, referred to as the "instance of Monad for type F" or "computation expression for F". By implementing this interface using coroutines and continuations, Arrow provides a specialization that enables readable direct async/await direct style code for any data type that provides an effect block with `bind` capabilities.
```

Comment:
I'd redo this paragraph, it feels like you need to know a bunch of concepts to appreciate it. I'd stick to implementation details, like in the rest of the article.

Reply:
Took the second half out just to clarify the terminology as to how people call these in Kotlin